mne/io/edf/edf.py

```diff
@@ -147,20 +166,234 @@ def __init__(
+        edf_info["blob"] = input_fname if _file_like(input_fname) else None
+
+        _validate_type(units, (str, None, dict), "units")
+        if units is None:
+            units = dict()
+        elif isinstance(units, str):
+            units = {ch_name: units for ch_name in info["ch_names"]}
+
+        for k, (this_ch, this_unit) in enumerate(orig_units.items()):
+            if this_ch not in units:
+                continue
+            if this_unit not in ("", units[this_ch]):
+                raise ValueError(
+                    f"Unit for channel {this_ch} is present in the file as "
+                    f"{repr(this_unit)}, cannot overwrite it with the units "
+                    f"argument {repr(units[this_ch])}."
+                )
+            if this_unit == "":
+                orig_units[this_ch] = units[this_ch]
+                ch_type = edf_info["ch_types"][k]
+                scaling = _get_scaling(ch_type.lower(), orig_units[this_ch])
+                edf_info["units"][k] /= scaling
+
+        # Raw attributes
+        last_samps = [edf_info["nsamples"] - 1]
+        super().__init__(
+            info,
             preload,
+            filenames=[_path_from_fname(input_fname)],
+            raw_extras=[edf_info],
+            last_samps=last_samps,
+            orig_format="int",
+            orig_units=orig_units,
+            verbose=verbose,
+        )
+
+        # Read annotations from file and set it
+        if len(edf_info["tal_idx"]) > 0:
+            # Read TAL data exploiting the header info (no regexp)
+            idx = np.empty(0, int)
+            tal_data = self._read_segment_file(
+                np.empty((0, self.n_times)),
+                idx,
+                0,
+                0,
+                int(self.n_times),
+                np.ones((len(idx), 1)),
+                None,
+            )
+            annotations = _read_annotations_edf(
+                tal_data[0],
+                ch_names=info["ch_names"],
+                encoding=encoding,
+            )
+            self.set_annotations(annotations, on_missing="warn")
+
+    def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        """Read a chunk of raw data."""
+        return _read_segment_file(
+            data,
+            idx,
+            fi,
+            start,
+            stop,
+            self._raw_extras[fi],
+            self.filenames[fi]
+            if self._raw_extras[fi]["blob"] is None
+            else self._raw_extras[fi]["blob"],
+            cals,
+            mult,
+        )
+
+
+def _path_from_fname(fname) -> Path | None:
+    if not isinstance(fname, Path):
+        if isinstance(fname, str):
+            fname = Path(fname)
+        else:
+            # Try to get a filename from the file-like object
+            try:
+                fname = Path(fname.name)
+            except Exception:
+                fname = None
+    return fname
```

Comment:
simplify logic? here it's 3-deep, but could be 1, e.g.

```suggestion
    if isinstance(fname, (str | Path)):
        return Path(fname)
    # Try to get a filename from the file-like object
    try:
        return Path(fname.name)
    except Exception:
        return None
```

Reply:
Fixed in edf.py, in `mne/io/fiff/raw.py` line 552 the same function is used however, I can change it if you will.